diag_manager/fms_doubly_linked_list.F90

```diff
@@ -0,0 +1,275 @@
+  ! newly inserted node.
+  function  insert_data( this, t_nd,  d ) result(liter)
+    class(linked_list), intent(in out) :: this
+    class(node), pointer, intent (in)  :: t_nd
+    class(*), target, intent(in)       :: d
+    class(literator), allocatable      :: liter
+    !!
+    class(node), pointer :: nd
+    allocate(nd)
+    nd%data => d
+    nd%prev => t_nd%prev
+    nd%next => t_nd
+    t_nd%prev%next => nd
+    t_nd%prev => nd
+    this%theSize = this%theSize + 1
+    liter = literator(nd, this%tail)
+  end function insert_data
+
+  !> Remove Node nd from the linked tree. Return the iterator
+  ! that begins with the nex node after nd, and ends with the
+  ! list end node. Return the list iterator if the node cannot
+  ! be removed.
+  function  remove_node( this, nd ) result( litr)
+    class(linked_list), intent(in out) :: this
+    type(node), pointer, intent(in out) :: nd
+    class(literator), ALLOCATABLE  :: litr
+    !Dont even try to remove the head and tail nodes!
+    if (.not. ( associated (this%head , nd ) .or. &
+      associated (this%tail , nd ) )) THEN
+      litr = literator( nd%next, this%tail )
+      nd%prev%next => nd%next
+      nd%next%prev => nd%prev
+      deallocate(nd)
+      this%theSize =  this%theSize - 1
+    else
+      litr = this%get_literator()
+    endif
+  end function remove_node
+
+
+  !>  Remove the data at the head of the list. Return an
+  !  Iterator starting with the next (new front) node, and
+  ! Return the list iterator if the list is already empty
+  function pop_at_front (this ) result( liter )
+    class(linked_list), intent(in out) :: this
+    class(literator), allocatable :: liter
+    !!
+    class(node), pointer :: nd
+    if(this%theSize /= 0) then
+      nd => this%head%next !!  TODO: verify
+      liter =  this%remove(nd)
+    else
+      liter = this%get_literator()
+    endif
+  end function pop_at_front
+
+  function pop_at_back (this ) result( liter )
+    class(linked_list), intent(in out) :: this
+    class(literator) , allocatable :: liter
+    !!
+    class(node), pointer :: nd
+    if(this%theSize /= 0) then
+      nd => this%tail%prev !! TODO: verify
+      liter = this%remove( nd )
+    else
+      liter = this%get_literator()
+    endif
+  end function pop_at_back
+
+
+
+  !> push (insert) data at the head of the list
+  function push_at_front( this, d ) result(litr)
+    class(linked_list), intent(in out) :: this
+    class(*), target, intent(in out) :: d
+    class(literator), allocatable :: litr
+    litr = this%insert (this%head%next, d)
+  end function push_at_front
+
+  !> push (insert) data at the end of the list
+  function push_at_back( this, d ) result(litr)
+    class(linked_list), intent(in out) :: this
+    class(*), target, intent(in out) :: d
+    class(literator), allocatable :: litr
+    litr =  this%insert (this%tail, d)
+  end function push_at_back
+
+  function node_constructor () result (nd)
```

Comment:
This function needs a doxygen description

Reply:
Done